### PR TITLE
Tuist plugins 2.0 fetch

### PR DESCRIPTION
**Pull request by fortmarek**
### Short description 📝

This PR does the heavy-lifting for the new plugins architecture. The remaining next steps are tracked [here](https://github.com/tuist/tuist/pull/3492)

It introduces a new command `tuist fetch` that will fetch both dependencies and plugins - the reason for this change was to make it easier to fetch all necessary content for the project. 

For fetching the remote plugins, the following happens:
- if the plugin is integrated with a SHA, nothing is done. Later, we might want to build the plugin but it is out-of-scope for the first iteration and only tagged commits will be considered
- if the plugin is integrated with a git tag, the repository is cloned (as was the case before). Afterwards, we download the release artifacts of a given tag (such as [here](https://github.com/fortmarek/ExampleTuistPlugin/releases/tag/0.2.0)), unzip it to a given directory in cache and change files with `tuist-` prefix to be executable. When `tuist xxx` is invoked and it is not one of tuist own commands, it will search through plugins' tasks and invoke the first result. If none is found, we try to find the task in the global env and if that also fails, we will throw an error saying that the command does not exist.

If you want to try this out, simply checkout this branch and run `tuist fetch` in `app_with_plugins`. Afterwards, you should be able to run `tuist create-file` which runs a custom task called `tuist-create-file` 🎉 

### Checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [ ] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [ ] In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

**Comment by luispadron** [comment]
Going to take a look at this soon, also left some thoughts here: https://github.com/tuist/tuist/discussions/3491#discussioncomment-1505315

File: Package.resolved [hunk 1]
```diff
@@ -28,22 +28,13 @@
           "version": "1.0.2"
         }
       },
-      {
```
[the remark this answers is not in the document]
Comment (fortmarek) [reply Package.resolved]: good catch, fixed!

File: Sources/TuistKit/Commands/TuistCommand.swift [hunk 2]
```diff
@@ -46,11 +47,10 @@ public struct TuistCommand: ParsableCommand {
         let executeCommand: () throws -> Void
         do {
             let processedArguments = Array(processArguments(arguments)?.dropFirst() ?? [])
-            let commandName = processedArguments.first ?? ""
             let isTuistCommand = Self.configuration.subcommands
-                .map { $0._commandName }
+                .map({ $0._commandName })
```
[the remark this answers is not in the document]
Comment (fortmarek) [reply Sources/TuistKit/Commands/TuistCommand.swift:51]: Unfortunately, the current form is the only one possible (I am getting `Key path cannot refer to static member '_commandName'` if I would apply your suggestion.

File: Sources/TuistKit/Commands/DependenciesCommand.swift [hunk 3]
```diff
@@ -8,7 +8,6 @@ struct DependenciesCommand: ParsableCommand {
             commandName: "dependencies",
             abstract: "A set of commands for dependencies' management.",
             subcommands: [
-                DependenciesFetchCommand.self,
                 DependenciesUpdateCommand.self,
                 DependenciesCleanCommand.self,
```
[the remark this answers is not in the document]
Comment (fortmarek) [reply Sources/TuistKit/Commands/DependenciesCommand.swift:12]: @danyf90 @luispadron @pepibumur 

I thought of flattening `tuist dependencies update` and `tuist dependencies clean` into `tuist clean` and `tuist update`. `tuist update` would, however, mean updating tuist version itself **and** dependencies (in the future also possibly plugins). It should be possible to do in code but I am not sure if users would expect this to happen. What do you think? Do you have a suggestion for a different name for the command `tuist update` that would mean updating all tuist artifacts (dependencies and plugins)?

**Pull request merged**